src/parser.rs

```diff
@@ -23,7 +23,23 @@ use crate::Identifier;
 pub const FULL_MATCH: &str = "__tsg__full_match";
 
 impl ast::File {
+    /// Parses a graph DSL file, returning a new `File` instance.
+    pub fn from_source(
+        language: Language,
+        ctx: &mut Context,
+        source: &str,
+    ) -> Result<Self, ParseError> {
+        let mut file = ast::File::new(language);
+        #[allow(deprecated)]
+        file.parse(ctx, source)?;
+        file.check(ctx)?;
+        Ok(file)
+    }
+
     /// Parses a graph DSL file, adding its content to an existing `File` instance.
+    #[deprecated(
+        note = "Parsing multiple times into the same `File` instance is unsound. Use `File::from_source` instead."
+    )]
```

Comment:
Interesting!  I had originally added this method in the hopes that we'd implement "include files" or something similar using it.  (Hand waving: you'd pass multiple `.tsg` files on the command line, and we'd call `parse` on each one in order to build up the combined TSG file before executing.)

There's definitely no need to tie ourselves in knots supporting that pattern, especially since we haven't actually implemented include files yet.  Whenever we do, we can change things around as needed at the API level.

But, the deprecation comment makes we wonder whether the unsoundness that you're describing will _preclude_ the possibility of include files?

Reply:
> I had originally added this method in the hopes that we'd implement "include files" or something similar using it.

Interesting idea! I hadn't thought of that use case.

I think this method was perfectly fine until I implemented query merging. Because I merge all queries into one at the end of a parse, it would overwrite an earlier merged query. If we want to support this, we must make sure we merge all queries from all files correctly. (If only we could incrementally add patterns to query...) Another question is if the computed capture indices in previous files remain valid in the new merged query. This might all work out nicely, or be annoying, but I'm saving that work for when actually looking at imports :)